### PR TITLE
New data type

### PR DESCRIPTION
**Pull request by Cpaulyz**
### Content3 ...

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

**Review by HTHou — COMMENTED** [review]
Hi, there are some code in TsFileProcessor I think need to be considered.
https://github.com/apache/iotdb/blob/b0306420a1079f47a6743087773ba1ae4517d610/iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/memtable/TsFileProcessor.java#L592

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/IoTDBDescriptor.java [hunk 1]
```diff
@@ -2234,8 +2234,10 @@ public TSEncoding getDefaultEncodingByType(TSDataType dataType) {
       case BOOLEAN:
         return conf.getDefaultBooleanEncoding();
       case INT32:
+      case DATE:
         return conf.getDefaultInt32Encoding();
       case INT64:
+      case STRING:
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/IoTDBDescriptor.java:2240]: It should be TIMESTAMP here. Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/rest/utils/InsertRowDataUtils.java [hunk 2]
```diff
@@ -95,6 +97,8 @@ public static List<Object> reGenValues(List<TSDataType> types, List<Object> valu
           }
           break;
         case TEXT:
+        case BLOB:
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/rest/utils/InsertRowDataUtils.java:100]: I have not used RestAPI for writing. I need to go through some examples or unit tests to help me with this part later.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/statement/literal/Literal.java [hunk 3]
```diff
@@ -36,7 +36,8 @@ public enum LiteralType {
     DOUBLE,
     LONG,
     STRING,
-    NULL
+    NULL,
+    BINARY
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/statement/literal/Literal.java:40]: Added. It seems that Literal doesn't get serialised and deserialised when writing with sql, so I didn't notice.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/statement/crud/InsertStatement.java [hunk 4]
```diff
@@ -43,7 +43,7 @@ public class InsertStatement extends Statement {
   private long[] times;
   private String[] measurementList;
 
-  private List<String[]> valuesList;
+  private List<Object[]> valuesList;
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/statement/crud/InsertStatement.java:46]: I forget why I modified this. But it looks like String[] is correct. I REVERT this modification.

**Comment by Cpaulyz** [comment]
> Hi, there are some code in TsFileProcessor I think need to be considered.
> 
> https://github.com/apache/iotdb/blob/b0306420a1079f47a6743087773ba1ae4517d610/iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/memtable/TsFileProcessor.java#L592

@HTHou Thanks for the reminder! I missed some checks on TEXT. I made some changes in commit 228bba7515. Please help review them for me~

I think adding an isBinary() method to TSDataType would have been a better choice, but that would have required to modify the TsFile repository.

File: iotdb-client/service-rpc/src/main/java/org/apache/iotdb/rpc/IoTDBJDBCDataSet.java [hunk 5]
```diff
@@ -573,7 +588,18 @@ public String getString(int index, TSDataType tsDataType, byte[][] values) {
...
+      case STRING:
         return new String(values[index], StandardCharsets.UTF_8);
+      case BLOB:
+        return BytesUtils.parseBlobByteArrayToString(values[index]);
+      case TIMESTAMP:
+        return RpcUtils.formatDatetime(
+            RpcUtils.DEFAULT_TIME_FORMAT,
+            "ms",
+            BytesUtils.bytesToLong(values[index]),
+            ZoneId.systemDefault());
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-client/service-rpc/src/main/java/org/apache/iotdb/rpc/IoTDBJDBCDataSet.java]: Added.

**Comment by HTHou** [comment]
> > Hi, there are some code in TsFileProcessor I think need to be considered.
> > https://github.com/apache/iotdb/blob/b0306420a1079f47a6743087773ba1ae4517d610/iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/memtable/TsFileProcessor.java#L592
> 
> @HTHou Thanks for the reminder! I missed some checks on TEXT. I made some changes in commit [228bba7](https://github.com/apache/iotdb/commit/228bba7515af6f83073483b8f047aeef79fa7bff). Please help review them for me~
> 
> I think adding an isBinary() method to TSDataType would have been a better choice, but that would have required to modify the TsFile repository.

+1

**Comment by Cpaulyz** [comment]
Hi all, I will close this PR and open a new one for cooperation: https://github.com/apache/iotdb/pull/12576

**Pull request closed without merging**